src/engine/sparqlExpressions/BlankNodeExpression.cpp

```diff
@@ -0,0 +1,106 @@
+
+#include <absl/strings/str_cat.h>
+
+#include "engine/sparqlExpressions/LiteralExpression.h"
+#include "engine/sparqlExpressions/NaryExpressionImpl.h"
+#include "engine/sparqlExpressions/SparqlExpressionValueGetters.h"
+#include "util/ChunkedForLoop.h"
+
+namespace sparqlExpression {
+
+namespace detail {
+
+// Turn a blank node string into an iri with a special prefix.
+ad_utility::triple_component::Iri createBlankNodeForLabel(
+    std::string_view label) {
+  AD_CONTRACT_CHECK(label.starts_with("_:"));
+  std::string iriString =
+      absl::StrCat(QLEVER_INTERNAL_BLANK_NODE_IRI_PREFIX, label, ">");
+  return ad_utility::triple_component::Iri::fromStringRepresentation(
+      std::move(iriString));
+}
+
+// Escape a string for use as a blank node label.
+std::string escapeStringForBlankNode(std::string_view input) {
+  std::string output;
+  for (char c : input) {
+    // TODO<RobinTF> Implement actual escaping algorithm.
+    if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')) {
+      output.push_back(c);
+    }
+  }
+  return output;
+}
+
+// Map a string to a blank node.
+[[maybe_unused]] auto stringToBlankNodeImpl =
+    [](std::optional<std::string> s) -> IdOrLiteralOrIri {
+  if (s.has_value()) {
+    std::string label =
+        BlankNode{false, escapeStringForBlankNode(std::move(s.value()))}
+            .toSparql();
+    return LiteralOrIri{createBlankNodeForLabel(label)};
+  } else {
+    return Id::makeUndefined();
+  }
+};
+
+// SparqlExpression representing the term "BNODE("something")".
+using SpecificBlankNodeExpression =
+    NARY<1, FV<decltype(stringToBlankNodeImpl), StringValueGetter>>;
+
+// SparqlExpression representing the term "BNODE()".
+class UniqueBlankNodeExpression : public SparqlExpression {
+  size_t label_ = 0;
+  mutable uint64_t cacheBreaker_ = 0;
+  mutable uint64_t counter_ = 0;
+
+ public:
+  // _____________________________________________________________________________
+  explicit UniqueBlankNodeExpression(size_t label) : label_{label} {}
+  UniqueBlankNodeExpression(const UniqueBlankNodeExpression&) = delete;
+  UniqueBlankNodeExpression& operator=(const UniqueBlankNodeExpression&) =
+      delete;
+
+  // _____________________________________________________________________________
+  ExpressionResult evaluate(EvaluationContext* context) const override {
+    std::string label{QLEVER_INTERNAL_BLANK_NODE_IRI_PREFIX};
+
+    VectorWithMemoryLimit<IdOrLiteralOrIri> result{context->_allocator};
+    const size_t numElements = context->_endIndex - context->_beginIndex;
+    result.reserve(numElements);
+
+    ad_utility::chunkedForLoop<1000>(
+        0, numElements,
+        [this, &result](size_t) {
+          auto uniqueLabel = absl::StrCat("_:bn", label_, "_", counter_);
+          counter_++;
+          result.push_back(LiteralOrIri{createBlankNodeForLabel(uniqueLabel)});
+        },
+        [context]() { context->cancellationHandle_->throwIfCancelled(); });
+    return result;
+  }
+
```

Comment:
Add a TODO that in a future version this definitely shouldn't be implemented using strings, but using IDs, but that this changes in the interface as we basically need to pass something like `vector<Id> + localBlankNodeManager`.